core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingInterceptor.java

```diff
@@ -140,119 +131,90 @@ public final Object visitGetCacheEntryCommand(InvocationContext ctx, GetCacheEnt
-               // get response.
+      return invokeNextAndFinally(ctx, command, handleDataReadReturn);
+   }
+
+   private void handleDataReadReturn(InvocationContext rCtx, VisitableCommand rCommand, Object rv, Throwable t) throws Throwable {
+      DataCommand dataCommand = (DataCommand) rCommand;
+      if (t != null) {
+         if (t instanceof RpcException && performPartitionCheck(rCtx, dataCommand)) {
+            // We must have received an AvailabilityException from one of the owners.
+            // There is no way to verify the cause here, but there isn't any other way to get an invalid
+            // get response.
+            throw log.degradedModeKeyUnavailable(dataCommand.getKey());
+         } else {
+            if (t instanceof AllOwnersLostException && performPartitionCheck(rCtx, dataCommand)) {
                throw log.degradedModeKeyUnavailable(dataCommand.getKey());
-            } else {
-               // If all owners left and we still haven't received the availability update yet,
-               // we get OutdatedTopologyException from BaseDistributionInterceptor.retrieveFromProperSource
-               if (t instanceof AllOwnersLostException && performPartitionCheck(rCtx, dataCommand)) {
-                  // Unlike in PartitionHandlingManager.checkRead(), here we ignore the availability status
-                  // and we only fail the operation if _all_ owners have left the cluster.
-                  // TODO Move this to the availability strategy when implementing ISPN-4624
-                  // TODO We should collect the owners and check the topology atomically
-                  Collection<Address> owners =
-                        distributionManager.getCacheTopology().getDistribution(dataCommand.getKey()).readOwners();
-                  CacheTopology cacheTopology = stateTransferManager.getCacheTopology();
-                  if (cacheTopology == null || cacheTopology.getTopologyId() != dataCommand.getTopologyId()) {
-                     // just rethrow the exception
-                     throw t;
-                  }
-                  if (!InfinispanCollections.containsAny(transport.getMembers(), owners)) {
-                     throw log.degradedModeKeyUnavailable(dataCommand.getKey());
-                  }
-               }
-               throw t;
             }
+            throw t;
          }
+      }
 
-         postOperationPartitionCheck(rCtx, dataCommand, dataCommand.getKey());
-      });
+      if (performPartitionCheck(rCtx, dataCommand)) {
+         // We do the availability check after the read, because the cache may have entered degraded mode
+         // while we were reading from a remote node.
+         partitionHandlingManager.checkRead(dataCommand.getKey());
+      }
+      // TODO We can still return a stale value if the other partition stayed active without us and we haven't entered degraded mode yet.
```

Comment:
The only way to fix that would be to always expect a response from a majority of owners, and I don't see us doing that in the foreseeable future, so please remove the `TODO` tag from the comment :)